File: milestone 2/python_modules/module01/ex5/ft_plant_types.py
import math


class Plant:
    def __init__(self, n, h, a):
        self.name = n
        self.height = h
        self.age = a

    def print_info(self):
        class_name = self.__class__.__name__
        print(f"{self.name} ({class_name}): {self.height} cm, ", end="")
        print(f"{self.age} days old ", end="")


class Flower(Plant):
    def __init__(self, n, h, a):
        super().__init__(n, h, a)
        self.color = "red"
        self.is_blooming = True

    def print_info(self):
        super().print_info()
        print(f"{self.color} color")
        self.check_blooming()

    def check_blooming(self):
        if self.is_blooming:
            print(f"{self.name} is blooming beautifully!")
        else:
            print(f"{self.name} is not ready yet!")


class Tree(Plant):
    def __init__(self, n, h, a):
        super().__init__(n, h, a)
        self.trunk_diameter = 50.0

    def produce_shade(self):
        shade_area = (self.trunk_diameter ** 2 * math.pi) / 4
        print(f"{self.name} provides {shade_area:.2f} square meters of shade")

    def print_info(self):
        super().print_info()
        print(f"{self.trunk_diameter} cm diameter")


class Vegetable(Plant):
    def __init__(self, n, h, a):
        super().__init__(n, h, a)
        self.harvest_season = "summer "
        self.nutritional_value = "is rich in vitamin C"

    def print_info(self):
        super().print_info()
        print(f"{self.harvest_season} harvest")

    def print_message(self):
        print(f"{self.name} {self.nutritional_value}")


if __name__ == "__main__":
    print("=== Garden Plant Types ===\n")

    rose = Flower("Rose", 25, 30)
    oak = Tree("Oak", 500, 1825)
    tomato = Vegetable("Tomato", 80, 90)

    rose.print_info()
    oak.print_info()
    oak.produce_shade()
    tomato.print_message()
    tomato.print_info()

File: milestone 2/python_modules/module01/ex5/ft_plant_types.cpp
#include <iostream>
#include <string>
#include <cmath>

class Plant
{
  public:
	std::string name;
	int height, age;

	Plant(std::string n, int h, int a) : name(n), height(h), age(a)
	{
	}
	virtual void PrintInfo()
	{
		std::string class_name;
		class_name = typeid(*this).name();
		class_name.erase(class_name.begin());
		std::cout << name << " (" << class_name << "): " << height << "cm, " << age << " days, ";
	}
};

class Flower : public Plant
{
public:
    Flower(std::string n, int h, int a) : Plant(n, h, a) {}

    std::string color = "red";

    void PrintInfo() override
    {
        Plant::PrintInfo();
        std::cout << color << " color" << std::endl;
    }

    bool bloom(bool isBlooming)
    {
        if (isBlooming)
            std::cout << name << " is blooming beautifully!" << std::endl;
        else
            std::cout << name << " is not blooming yet!" << std::endl;
        return isBlooming;
    }
};


class Tree : public Plant
{
  public:
	Tree(std::string n, int h, int a) : Plant(n, h, a)
	{
	}
	float trunk_diameter = 50;
	void produce_shade()
	{
		std::cout << name << " provides " << (std::pow(trunk_diameter, 2) * M_PI)/ 4;
		std::cout << " square meters of shade" << "\n";
	}
	void PrintInfo() override
	{
		Plant::PrintInfo();
		std::cout << trunk_diameter << "cm diameter" << std::endl;
	}
};

class Vegetable : public Plant
{
  public:
	Vegetable(std::string n, int h, int a) : Plant(n, h, a)
	{
	}
	std::string harvest_season = "summer ";
	std::string nutritional_value = " is rich in vitamin C";
	void PrintInfo() override
	{
		Plant::PrintInfo();
		std::cout << harvest_season << "harvest" << std::endl;
	}
	void print_message()
	{
		std::cout << name << nutritional_value << std::endl; 
	}
};

int	main(void)
{
	std::cout << "=== Garden Plant Types ===" << "\n\n";
	Flower rose("Rose", 25, 30);
	Tree	oak("Oak", 500, 1825);
	Vegetable tomato("Tomato", 80, 90);

	rose.PrintInfo();
	rose.bloom(true);
	std::cout << "\n";
	oak.PrintInfo();
	oak.produce_shade();
	std::cout << "\n";
	tomato.PrintInfo();
	tomato.print_message();

	std::cout << "\n";
	return (0);
}